File: swift/cmdline/MazesForProgrammers/MazesForProgrammers/cell.swift
//
//  cell.swift
//  MazesForProgrammers
//
//  Created by Martin Linklater on 28/12/2023.
//

import Foundation

class Cell {
    var row: Int = -1
    var col: Int = -1
    var north: Cell?
    var south: Cell?
    var east: Cell?
    var west: Cell?
    var links: [Cell]
    
    init(row: Int, col: Int) {
        self.row = row
        self.col = col
    }
}

File: ruby/cell.rb
class Cell
    attr_reader :row, :column
    attr_accessor :north, :south, :east, :west

    def initialize(row, column)
        @row = row
        @column = column
        @links = {}
    end

    # Links a cell with another cell. If bidi is true, the link is bidirectional.
    def link(cell, bidi=true)
        @links[cell] = true
        cell.link(self, false) if bidi
        self
    end

    # Unlinks a cell from another cell. If bidi is true, the unlink is bidirectional.
    def unlink(cell, bidi=true)
        @links.delete(cell)
        cell.unlink(self, false) if bidi
        self
    end

    def links
        @links.keys
    end

    def linked?(cell)
        @links.key?(cell)
    end

    def neighbors
        list = []
        list << north if north
        list << south if south
        list << east if east
        list << west if west
        list
    end
end